api/store.proto

```diff
@@ -0,0 +1,185 @@
+		option (docker.protobuf.plugin.tls_authorization) = { roles: "swarm-manager" };
+	};
+
+	// Watch starts a stream that returns any changes to objects that match
+	// the specified selectors. When the stream begins, it immediately sends
+	// an empty message back to the client. It is important to wait for
+	// this message before taking any actions that depend on an established
+	// stream of changes for consistency.
+	rpc Watch(WatchRequest) returns (stream WatchMessage) {
+		option (docker.protobuf.plugin.tls_authorization) = { roles: "swarm-manager" };
+	};
+
+	// View creates a snapshot of the current state of the store, and
+	// accepts a stream of queries against it. Each query returns a
+	// response in the response stream.
+	// TODO(aaronl): Does it make sense to support this in the API?
+	//rpc View(stream ViewRequest) returns (stream ViewResponse) {
+	//	option (docker.protobuf.plugin.tls_authorization) = { roles: "swarm-manager" };
+	//}
+}
+
+message GetObjectRequest {
+	// Kind can contain a builtin type such as "node", "secret", etc. or
+	// the kind specified by a custom-defined object.
+	string kind = 1;
+
+	// ID of the object to retrieve.
+	string object_id = 2 [(gogoproto.customname) = "ObjectID"];
+}
+
+message FindObjectsRequest {
+	// Kind can contain a builtin type such as "node", "secret", etc. or
+	// the kind specified by a custom-defined object.
+	string kind = 1;
+
+	// Selectors are combined with OR logic. An empty list of selectors
+	// causes all objects of the given type to be returned.
+	repeated SelectBy selectors = 2;
+}
+
+message FindObjectsResponse {
+	repeated Object results = 1;
+}
+
+message UpdateObjectsRequest {
+	repeated Object objects = 1;
+}
+
+message UpdateObjectsResponse {
+	// Objects contains the updated objects. On a successful return, these
+	// will be the same as the objects passed in UpdateObjectsRequest,
+	// except with Meta updated.
+	repeated Object objects = 1;
+
+	// TODO(aaronl): Provide structured information about which object
+	// couldn't be updated, and why, in error cases?
+}
+
+message WatchRequest {
+	message WatchEntry {
+		// Kind can contain a builtin type such as "node", "secret", etc. or
+		// the kind specified by a custom-defined object.
+		string kind = 1;
+
+		// Action (create/update/delete)
+		StoreActionKind action = 2;
+
+		// Filters are combined using AND logic - an event must match
+		// all of them to pass the filter.
+		repeated SelectBy filters = 3;
+	}
+
+	// Multiple entries are combined using OR logic - i.e. if an event
+	// matches all of the selectors specified in any single watch entry,
+	// the event will be sent to the client.
+	repeated WatchEntry entries = 1;
+}
+
+// WatchMessage is the type of the stream that's returned to the client by
+// Watch. Note that the first item of this stream will always be a WatchMessage
+// with a nil Object, to signal that the stream has started.
+message WatchMessage {
```

Comment:
`kind` as in create/update/delete or node/service/task/etc?

I don't see how either would work. Each `WatchMessage` may contain many events. For example if you watch the entire event stream, you'll see some transactions that do several things. These events must say batched in the event stream, because they have the same `version`.